webrender/src/resource_cache.rs

```diff
@@ -606,41 +668,75 @@ impl ResourceCache {
             // The image or tiling size is too big for hardware texture size.
             warn!("Dropping image, image:(w:{},h:{}, tile:{}) is too big for hardware!",
                   template.descriptor.size.width, template.descriptor.size.height, template.tiling.unwrap_or(0));
-            self.cached_images.insert(request, Err(ImageCacheError::OverLimitSize));
+            self.cached_images.insert(request.key, ImageResult::Err(ImageCacheError::OverLimitSize));
             return;
         }
 
-        // If this image exists in the texture cache, *and* the epoch
-        // in the cache matches that of the template, then it is
-        // valid to use as-is.
-        let (entry, needs_update) = match self.cached_images.entry(request) {
-            Occupied(entry) => {
-                let info = entry.into_mut();
-                let needs_update = info.as_mut().unwrap().epoch != template.epoch;
-                info.as_mut().unwrap().epoch = template.epoch;
-                (info, needs_update)
+        let storage = match self.cached_images.entry(request.key) {
+            Occupied(e) => {
```

Comment:
perhaps, it would be cleaner to lift the condition up here:
```rust
Occupied(e) if e.is_untiled_auto() => {
...
}
Occupied(e) => {
...
}
Vacant(e) => {..}
```